js/web/lib/wasm/jsep/util.ts

```diff
@@ -0,0 +1,766 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+/* eslint-disable no-param-reassign */
+
+export class MatMulUtil {
+  /**
+   * Fix the input shapes for MatMul operation if they need fixing
+   * @param dimsA The shape of tensor A. Should be an array of positive integers
+   * @param dimsB The shape of tensor B. Should be an array of positive integers
+   * @returns A tuple containing the preprocessed input shapes as required by ONNX specifications
+   */
+  static preprocessInputShapes(dimsA: readonly number[], dimsB: readonly number[]):
```

Comment:
in future we may or may not shift to C/C++ implementation, depending on whether the core kernel logic will be implemented in TS or C/C++. passing values between languages are high-cost, especially when asyncify is enabled.

Reply:
if you only support fixed size values (other comment said symbolic dims were not supported) this can be one off call though so runtime cost is minimized. the development and maintenance cost of this EP is significantly higher when we can't use an existing implementation of common code.